pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -370,10 +466,77 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 	}
 	var err error
 	for k, v := range configs {
-		resources, err = addConfigMapItem(master, getConfigName(master, k), labels, v, resources)
+		resources, err = addConfigMapItem(master, getResourceName(master, k), labels, v, resources)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If upgrade has failed, do not set imageToUse, just return
+	if master.Status.Meta.IsConditionTrue(upgradeFailed) {
+		// User has reset the CR after failure. Remove upgrade failed status.
+		if master.Spec.Image == master.Status.ImageToUse {
+			master.Status.ClearCondition(upgradeFailed, upgradeResetMessage, upgradeResetMessage)
+		}
+		return resources, nil
+	}
+
+	// Early return if image does not need updating. Update imageToUse since we are in a non
+	// failure case.
+	if master.Status.ImageToUse == "" || master.Status.ImageToUse == master.Spec.Image {
+		master.Status.ImageToUse = master.Spec.Image
+		master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
+		return resources, nil
+	}
+
+	var upgradeResources []reconciler.Object
+	if master.Status.UpgradeJobVersion == "" {
```

Comment:
You can flip the condition to reduce code duplication in this if block and the `upgradeJob == nil` case.

```java
var upgradeJob *batchv1.Job
if master.Status.UpgradeJobVersion != "" {
	item := k8s.GetItem(observed, &batchv1.Job{}, getUpgradeJobName(master), "default")
	if item != nil {
		upgradeJob = item.(*batchv1.Job)
	}
}
if upgradeJob == nil {
		// Add a new upgrade job, and set ready status to false
		var err error
		upgradeResources, err = prepareNewPreUpgradeJobResources(master, rsrclabels)
		if err != nil {
			return nil, err
		}
		addUpgradeComponentNotReady(master)
		return append(resources, upgradeResources...), nil
}
// Otherwise upgradeJob exists, handle succeed / failure case.
```

Reply:
oh nice!